File: getting-started/golang/merge_sort.go
package main

import (
	"fmt"
	"math/rand"
)

func sort(data []int) {
	mergeSort(data, 0, len(data)-1)
}

func mergeSort(data []int, startIdx, endIdx int) {
	if startIdx < endIdx {
		middleIdx := (startIdx + endIdx) / 2

		mergeSort(data, startIdx, middleIdx)
		mergeSort(data, middleIdx+1, endIdx)

		merge(data, startIdx, middleIdx, endIdx)
	}
}

func merge(data []int, startIdx, middleIdx, endIdx int) {
	firstHalf := append([]int{}, data[startIdx:middleIdx+1]...)
	secondHalf := append([]int{}, data[middleIdx+1:endIdx+1]...)

	i, j := 0, 0
	for k := startIdx; k < endIdx+1; k++ {
		if i < len(firstHalf) && j < len(secondHalf) {
			if firstHalf[i] < secondHalf[j] {
				data[k] = firstHalf[i]
				i++
			} else {
				data[k] = secondHalf[j]
				j++
			}
		} else if i < len(firstHalf) {
			data[k] = firstHalf[i]
			i++
		} else {
			data[k] = secondHalf[j]
			j++
		}
	}

	fmt.Println(data)
}

func main() {
	data := make([]int, 20)

	for i := range 20 {
		data[i] = rand.Intn(20) + 1
	}

	fmt.Println("--- before ---")
	fmt.Println(data)

	fmt.Println("--- sort ---")
	sort(data)

	fmt.Println("--- after ---")
	fmt.Println(data)
}

File: getting-started/python/merge_sort.py
import random


def sort(data):
    merge_sort(data, 0, len(data) - 1)


def merge_sort(data, start_idx, end_index):
    if start_idx < end_index:
        middle_idx = (start_idx + end_index) // 2

        merge_sort(data, start_idx, middle_idx)
        merge_sort(data, middle_idx + 1, end_index)

        merge(data, start_idx, middle_idx, end_index)


def merge(data, start_idx, middle_idx, end_index):
    first_half = data[start_idx : middle_idx + 1]
    second_half = data[middle_idx + 1 : end_index + 1]

    i = j = 0
    for k in range(start_idx, end_index + 1):
        if i < len(first_half) and j < len(second_half):
            if first_half[i] < second_half[j]:
                data[k] = first_half[i]
                i += 1
            else:
                data[k] = second_half[j]
                j += 1
        elif i < len(first_half):
            data[k] = first_half[i]
            i += 1
        else:
            data[k] = second_half[j]
            j += 1

    print(data)


if __name__ == "__main__":
    data = [random.randint(1, 20) for _ in range(20)]

    print("--- before ---")
    print(data)

    print("--- sort ---")
    sort(data)

    print("--- after ---")
    print(data)
